### PR TITLE
Add balance form for easy inventory

### PR DESCRIPTION
**Pull request by blaggacao**

### CODE REVIEW

File: src/app/balance-form/balance-form.component.html [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    [class.field-error]="hasError('amount')"
+  >
+    <TextField
+      id="amount-field"
+      class="field"
+      formControlName="amount"
+      [keyboardType]="amountFieldKeyboardType"
+      (loaded)="onAmountFieldLoaded($event)"
+      (returnPress)="hideKeyboard()"
+      (doubleTap)="enableArithmeticExpressions()"
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/balance-form/balance-form.component.html:92]: (7*4)+2
7 x packs-of-4-apples + 2apples

File: src/app/balance-form/balance-form.component.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#commodity-field {
+    font-size: 28;
+}
+
+#amount-field {
+    font-size: 28;
+}
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/balance-form/balance-form.component.scss:7]: The commodity is the _main_ field here. It is equally important as the amount.

The semantic difference by making this same-sized is actually intended.

Doing an inventory without a unit would be fun. :smile:

File: src/app/commodity-form/commodity-form.component.html [hunk 3]
```diff
@@ -5,7 +5,7 @@
     text="Back"
     (tap)="goBack()"
   ></NavigationButton>
-  <Label class="action-bar-title" text="New currency"></Label>
+  <Label class="action-bar-title" text="New commodity"></Label>
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/commodity-form/commodity-form.component.html:8]: While the beancount docs use these interchangeably, when mapping to real world use cases, it becomes clear that commodity is the more generic one:

- Please enter the currency: USD - OK
- Please enter the currency: WALNUTS - NOK

- Please enter the commodity: USD - OK
- Please enter the commodity: WALNUTS - OK

File: src/app/shared/balance.model.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+//         this.date = new Date(options.date)
+//         this.account = options.account
+//         this.accountPad = options.accountPad
+//     }
+//
+//     toBeancount(): string {
+//         const dateStr = getDateStr(this.date)
+//         return `${dateStr} pad ${this.account} ${this.accountPad}\n`
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/shared/balance.model.ts:41]: Yes, in fact, a padded inventory is one where you allow differences to be written off.

For example, if there are missing two sacks of potatoes, with a padded inventory, they would be written off to a specified account.

I would even say, that without padding the inventory ("balance") function is only half as usable.

The real problem to implement this cleanly is that currently padding-per-commodity is not  supported.

Once that is supported, I plan on doing a flag similar to the tx amount flag, but using `[ = , ~ ]`.

But I'd have to wait for that to be implemented upstream.

File: src/app/balance-form/balance-form.component.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/balance-form/balance-form.component.scss:7]: Think about how you'd do an inventory on your mobile:

1. enter the shelve ("account")
2. select the commodity to count -> `CHICKENWINGS`
3. count the chicken wings ("amount")

File: src/app/commodity-form/commodity-form.component.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/commodity-form/commodity-form.component.html:8]: I mean, it's not 100% optimal either way. But at least my users would be just a tiny bit _less_ confused.

File: src/app/shared/balance.model.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/app/shared/balance.model.ts:41]: Imagine for example a restaurant where the manager count's stock of spare-ribs. according to sales, 20 spare-ribs went over the counter, but the rats ate to more :smile:

**Comment by blaggacao** [comment]
@xuhcc Thanks a lot for your review/consideration!

I have to add I actually tested this on my phone after a good 5 hours of setting the environment up. :smile: 

My wife has a restaurant where we want to use beancount mobile to have the manager do the daily inventory which we then can cross with sales accordingly.

The workflow is that we share the `inventory.beancount` file only over syncthing with the manager so that they provide constant feed of inventories.

At the end of the day we review the inventories and when they are approved, just do a git commit of the additions.

_... managing a restaurant with gitops_ :rofl:

File: src/app/balance-form/balance-form.component.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (xuhcc) [reply src/app/balance-form/balance-form.component.scss:7]: Yes, all these fields are equally important and are required fields. But `font-size: 28` style is not used for required fields anywhere in the app, it only used for Amount field on Transaction page which is a special case.

File: src/app/commodity-form/commodity-form.component.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (xuhcc) [reply src/app/commodity-form/commodity-form.component.html:8]: This app is designed to track personal finances, hence the use of the word "currency".

If your use case is different, you can make a custom build for your users.

File: src/app/shared/balance.model.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (xuhcc) [reply src/app/shared/balance.model.ts:41]: OK, let's delete the commented lines for now

**Comment by xuhcc** [comment]
Thanks for your contribution!
I use it for personal accounting, but the workflow is similar: the beancount file is syncing across my devices, I regularly review it and commit changes to git.

**Comment by TRSx80** [comment]
Thanks for your efforts, @blaggacao!

Am I understanding correctly that this will implement a method to add balance directives?

**Comment by blaggacao** [comment]
Yes, however, In my use case, a private deployment to a couple of mobile phones makes most sense which is why I personally don't have use for the suggested renames. If somebody would pick this PR up and tend to the maintainer's comments, I'd more than welcome it. :smile:

**Comment by xuhcc** [comment]
@blaggacao 
I took your PR and adjusted a bit: #16. It is merged now.
Thanks!

**Comment by TRSx80** [comment]
Thanks guise!  I will give it a try!

I have been making note of cash balance in todo.txt with timestamps and refiling them later when at my computer.  lol  My main use for the mobile app is keeping track of cash transactions in fact (where balance assertions are important), so this should really streamline that workflow.

**Pull request closed without merging**